data_structures/hashing/bloom_filter.py

```diff
@@ -0,0 +1,105 @@
+"""
+See https://en.wikipedia.org/wiki/Bloom_filter
+"""
+from hashlib import md5, sha256
+from random import choices
+from string import ascii_lowercase
+
+
+class Bloom:
+    # number of hash functions is fixed
+    HASH_FUNCTIONS = (sha256, md5)
+
+    def __init__(self, size: int = 8) -> None:
+        self.bitstring = 0b0
+        self.size = size
+
+    def add(self, value: str) -> None:
+        h = self.hash_(value)
+        self.bitstring |= h
+        print(
+            f"""\
+[add] value =      {value}
+      hash =       {self.format_bin(h)}
+      filter =     {self.format_bin(self.bitstring)}
+"""
+        )
+
+    def exists(self, value: str) -> bool:
+        h = self.hash_(value)
+        res = (h & self.bitstring) == h
+
+        print(
+            f"""\
+[exists] value =   {value}
+         hash =    {self.format_bin(h)}
+         filter =  {self.format_bin(self.bitstring)}
+         res =     {res}
+"""
+        )
+        return res
+
+    def format_bin(self, value: int) -> str:
+        res = bin(value)[2:]
+        return res.zfill(self.size)
+
+    def hash_(self, value: str) -> int:
+        res = 0b0
+        for func in self.HASH_FUNCTIONS:
+            b = func(value.encode()).digest()
+            position = int.from_bytes(b, "little") % self.size
+            res |= 2**position
+        return res
+
+
+def test_movies() -> None:
+    b = Bloom()
+    b.add("Titanic")
+    b.add("Avatar")
+
+    assert b.exists("Titanic")
+    assert b.exists("Avatar")
+
+    assert b.exists("The Goodfather") in (True, False)
+    assert b.exists("Interstellar") in (True, False)
+    assert b.exists("Parasite") in (True, False)
+    assert b.exists("Pulp Fiction") in (True, False)
```

Comment:
We use [doctest](https://docs.python.org/3/library/doctest.html) to test our modules via a workflow

Reply:
Moved to module docstring